File: posts/cursos/matematica/calculo/code/r/taxa_variacao_intro.R
# arquivo: posts/cursos/matematica/calculo/code/r/taxa_variacao_intro.R
# Aproximação de inclinações de secantes e visualização básica

secant_slopes <- function(f, x0, hs) {
  ms <- sapply(hs, function(h) (f(x0 + h) - f(x0)) / h)
  data <- data.frame(h = hs, m_sec = ms)
  rownames(data) <- NULL
  return(data)
}

plot_secants <- function(f, x0, hs, window = c(x0 - 3, x0 + 3)) {
  xs <- seq(window[1], window[2], length.out = 400)
  ys <- sapply(xs, f)

  plot(xs, ys, type = "l", lwd = 2, xlab = "x", ylab = "f(x)")
  abline(v = x0, lty = 3)

  h_big  <- hs[1]
  h_small <- hs[length(hs)]

  m_big  <- (f(x0 + h_big)  - f(x0)) / h_big
  m_small <- (f(x0 + h_small) - f(x0)) / h_small

  abline(a = (f(x0) - m_big * x0), b = m_big, col = "gray50", lwd = 2)
  abline(a = (f(x0) - m_small * x0), b = m_small, col = "gray20", lwd = 2)

  legend("topleft",
         legend = c(paste0("secante (h=", h_big, ")"),
                    paste0("secante (h=", h_small, ")")),
         lty = 1, col = c("gray50", "gray20"), bty = "n")
}

File: posts/cursos/matematica/calculo/code/python/taxa_variacao_intro.py
# arquivo: posts/cursos/matematica/calculo/code/python/taxa_variacao_intro.py
import numpy as np
import matplotlib.pyplot as plt

def secant_slopes(f, x0, hs):
    """
    Retorna um array Nx2 com colunas [h, m_sec(h)],
    onde m_sec(h) = (f(x0+h) - f(x0)) / h.
    """
    hs = np.array(hs, dtype=float)
    msecs = np.array([(f(x0 + h) - f(x0)) / h for h in hs], dtype=float)
    return np.column_stack([hs, msecs])

def _line_through_point_slope(x, x0, y0, m):
    return m * (x - x0) + y0

def plot_secants(f, x0, hs, window=None):
    """
    Plota f(x) e duas secantes (uma com h maior e outra com h menor).
    """
    if window is None:
        window = (x0 - 3, x0 + 3)

    xs = np.linspace(window[0], window[1], 400)
    ys = np.vectorize(f)(xs)

    fig, ax = plt.subplots(figsize=(7.5, 4.5))
    ax.plot(xs, ys, linewidth=2, label="f(x)")
    ax.axvline(x0, linestyle="--", linewidth=1)

    h_big  = float(hs[0])
    h_small = float(hs[-1])

    m_big   = (f(x0 + h_big)   - f(x0)) / h_big
    m_small = (f(x0 + h_small) - f(x0)) / h_small

    y0 = f(x0)
    ax.plot(xs, _line_through_point_slope(xs, x0, y0, m_big),
            linewidth=2, alpha=0.7, label=f"secante (h={h_big})")
    ax.plot(xs, _line_through_point_slope(xs, x0, y0, m_small),
            linewidth=2, alpha=0.85, label=f"secante (h={h_small})")

    ax.set_xlabel("x")
    ax.set_ylabel("f(x)")
    ax.set_title("Secantes aproximando a tangente")
    ax.grid(True, alpha=0.3)
    ax.legend()
    plt.tight_layout()
    plt.show()

